src/tests/Microsoft.PowerFx.Core.Tests/ValueTests/ValueTests.cs

```diff
@@ -0,0 +1,325 @@
+
+            public double numberDouble { get; set; }
+
+            public bool boolean { get; set; }
+
+            public string str { get; set; }
+
+            public DateTime datetime { get; set; }
+
+            public DateTimeOffset dto { get; set; }
+
+            public TimeSpan timespan { get; set; }
+        }
+
+        [Fact]
+        public void RecordNotMarshalled()
+        {
+            RecordValue r = FormulaValue.NewRecord(new RowDontMarshal());
+            Assert.Equal(0, r.Fields.Count());
+        }
+
+        // These member kinds are not marshalled.  
+        private class RowDontMarshal
+        {
+            public static int StaticProp { get; set; }
+
+            private int PrivateProp { get; set; }
+
+            internal int InternalProp { get; set; }
+
+            public int publicField;
+        }
+
+        private class TestRow
+        {
+            public double a { get; set; }
+
+            public string str { get; set; }
+        }
+
+        [Fact]
+        public void Table()
+        {
+            var cache = new TypeMarshallerCache();
+            TableValue val = FormulaValue.NewTable(
+                cache,
+                new TestRow { a = 10, str = "alpha" },
+                new TestRow { a = 15, str = "beta" });
+
+            var field1 = ((StringValue)((RecordValue)val.Index(1).Value).GetField("str")).Value;
+            Assert.Equal("beta", field1);
+
+            dynamic d = val.ToObject();
+            Assert.Equal(15.0, d[1].a);
+
+            // Verify runtime json
+            var resultStr = val.Dump();
+
+            Assert.Equal("Table({a:10,str:\"alpha\"},{a:15,str:\"beta\"})", resultStr);
+        }
+                
+        [Fact]
+        public void TableFromRecords()
+        {
+            RecordValue r1 = FormulaValue.NewRecord(new TestRow { a = 10, str = "alpha" });
+            RecordValue r2 = FormulaValue.NewRecord(new TestRow { a = 15, str = "beta" });
+            TableValue val = FormulaValue.NewTable(r1, r2);
+
+            var result1 = ((RecordValue)val.Index(1).Value).GetField("a").ToObject();
+            Assert.Equal(15.0, result1);
+                        
+            dynamic d = val.ToObject();
+            Assert.Equal(10.0, d[0].a);
+
+            // Verify runtime json
+            var resultStr = val.Dump();
+            Assert.Equal("Table({a:10,str:\"alpha\"},{a:15,str:\"beta\"})", resultStr);
+
+            TableValue val2 = NewTableT(r1, r2);
+            Assert.Equal(resultStr, val2.Dump());
+        }
+
+        // Heterogenous table.
+        [Fact]
+        public void TableFromMixedRecords()
+        {
+            var cache = new TypeMarshallerCache();
+            RecordValue r1 = FormulaValue.NewRecord(new { a = 10, b = 20, c = 30 }, cache);
+            RecordValue r2 = FormulaValue.NewRecord(new { a = 11,         c = 31 }, cache);
+            TableValue val = FormulaValue.NewTable(r1, r2);
```

Comment:
Would it have the same result if the order was switched (i.e., first record containing only 'a' and 'c', and second record adding 'b')?

Reply:
No, it's pulling the first .... which could get very confusing ... I think we should require NewTable() takes in the recordType just to be explicit.